src/main/java/com/direwolf20/buildinggadgets/common/schema/template/provider/TemplateWorldSave.java

```diff
@@ -0,0 +1,108 @@
+
+    public TemplateWorldSave() {
+        super(NAME);
+        this.templates = new HashMap<>();
+    }
+
+    @Override
+    public void read(CompoundNBT nbt) {
+        if (! nbt.contains(KEY_TEMPLATE_LIST, NBT.TAG_LIST)) {
+            BuildingGadgets.LOGGER.warn("No templates found to load from nbt data.");
+            return;
+        }
+
+        ListNBT templates = nbt.getList(KEY_TEMPLATE_LIST, NBT.TAG_COMPOUND);
+        BuildingGadgets.LOGGER.debug("Loading {} template{}.", templates.size(), getSExtension(templates.size()));
+
+        int loadCount = 0;
+        for (INBT templateNBT : templates) {
+            CompoundNBT templateCompound = (CompoundNBT) templateNBT;
+            UUID id = templateCompound.getUniqueId(KEY_ID);
+            
+            boolean deserializeError = Template.deserializeNBT(templateCompound.getCompound(KEY_DATA))
+                    .map(t -> {
+                        this.templates.put(id, t);
+                        return Boolean.FALSE;
+                    })
+                    .orElse(Boolean.TRUE);
+
+            if (deserializeError)
+                BuildingGadgets.LOGGER.error("Could not load Template with id {}. Discarding.", id);
+            else
+                loadCount++;
+        }
+
+        if (loadCount == templates.size())
+            BuildingGadgets.LOGGER.debug("Successfully loaded {}/{} template{}.",
+                    templates.size(), templates.size(), getSExtension(templates.size()));
+        else
+            BuildingGadgets.LOGGER.warn("Loaded {}/{} template{} successfully. {} template{} could not be loaded!",
+                    loadCount, templates.size(), getSExtension(templates.size()), templates.size() - loadCount,
+                    getSExtension(templates.size() - loadCount));
+    }
+
+    @Override
+    public CompoundNBT write(CompoundNBT compound) {
+        BuildingGadgets.LOGGER.debug("Saving {} template{}.", templates.size(), getSExtension(templates.size()));
+
+        ListNBT templateList = new ListNBT();
+        for (Map.Entry<UUID, Template> entry : templates.entrySet()) {
+            CompoundNBT templateCompound = new CompoundNBT();
+            templateCompound.putUniqueId(KEY_ID, entry.getKey());
+            templateCompound.put(KEY_DATA, entry.getValue().serializeNBT());
+            templateList.add(templateCompound);
+        }
+
+        compound.put(KEY_TEMPLATE_LIST, templateList);
+
+        return compound;
+    }
+
+    public Optional<Template> getTemplate(UUID id) {//Templates are immutable => no markDirty
+        return Optional.ofNullable(templates.get(id));
+    }
+
+    public void setTemplate(UUID id, @Nullable Template template) {
+        Template current = templates.get(id); //avoid mark dirties...
+        if (current == template)
+            return;
+
+        if (template != null)
+            templates.put(id, template);
+        else
+            templates.remove(id);
```

Comment:
Please use seperate methods, I do not  like methods like this.

Reply:
Doing that would just move this check into the ServerTemplateProvider. I could change the ITemplateProvider interface though, to have a non-null set Method and a remove Method, which actually execute different code. I originally did not do that, as it adds some complexity everywhere, where we want to apply a Template which might or might not be present. Like for example when receiving a packet. It would also produce a few more Methods and I'm pretty sure you don't like it, if I split this one and the implementations of the providers into 3 Methods.